### PR TITLE
[trajopt] Add `MultipleShooting::AddConstraintToAllKnotPoints` variants

### PR DESCRIPTION
**Pull request by RussTedrake**
This significantly improves the ergonomics in pydrake for adding, for
instance, bounding box constraints to the state or input at all knot
points.  (And it resolves a long-standing TODO in
`multiple_shooting.h`).

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16847)

### CODE REVIEW

**Comment by RussTedrake** [comment]
+@hongkai-dai for feature review please?
cc @aykut-tri

File: systems/trajectory_optimization/multiple_shooting.h [hunk 1]
```diff
@@ -282,31 +282,50 @@ class MultipleShooting {
...
+    }
+    return bindings;
+  }
+
   /// Adds a constraint to all breakpoints, where any instances of time(),
   /// state(), and/or input() placeholder variables, as well as placeholder
   /// variables returned by calls to NewSequentialVariable(), are substituted
   /// with the relevant variables for each time index.
-  /// @return A vector of the constraints added to each knot point.
+  /// @return A vector of the bindings added to each knot point.
```
Comment (aykut-tri) [inline systems/trajectory_optimization/multiple_shooting.h:307]: Should this change entail renaming the `constraints` variable below for consistency?

**Review by aykut-tri — APPROVED** [review]
Looks good to me, only a super-minor comment.

**Review by hongkai-dai — COMMENTED** [review]
:lgtm: +@xuchenhan-tri for platform review please, thanks!

> Reviewed 8 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16847#-:-MysSXv42fPd3i6qvC-8:bq10k5o)* status: 2 unresolved discussions, LGTM missing from assignee xuchenhan-tri(platform), missing label for release notes (waiting on @hongkai-dai, @RussTedrake, and @xuchenhan-tri)

___
*[systems/trajectory_optimization/sequential_expression_manager.h, line 119 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16847#-MysO80w9_Yw9yAMoByy:-MysO80w9_Yw9yAMoByz:b-grkzha) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa8fac1023549297303867c09c82e6095865289/systems/trajectory_optimization/sequential_expression_manager.h#L119)):*
> ```objc
>       name_to_placeholders_;
>   std::unordered_map<symbolic::Variable, RowVectorX<symbolic::Expression>>
>       expressions_;
> ```

The name `expressions_` isn't clear enough. The reader might think this just contains the sequential expression, but actually it is a map from the placeholder variable to the expression for substitution. Shall we rename it as `placeholders_to_expressions_`?

**Review by xuchenhan-tri — COMMENTED** [review]
platform :lgtm: with one documentation request and a few nits.

> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16847#-:-MysjZNh2mUVV7Yu7fIC:b-dfn2ez)* status: 5 unresolved discussions, missing label for release notes (waiting on @RussTedrake)

___
*[systems/trajectory_optimization/multiple_shooting.h, line 296 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16847#-MysX2NvBb_VzGJttCCU:-MysX2Nw4amLW59O6wkH:b-cuavlo) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa8fac1023549297303867c09c82e6095865289/systems/trajectory_optimization/multiple_shooting.h#L296)):*
> ```objc
>       const Eigen::Ref<const VectorX<symbolic::Variable>>& vars) {
>     std::vector<solvers::Binding<C>> bindings;
>     for (int i = 0; i < N_; i++) {
> ```

nit, ditto throughout the pr.



_Suggestion:_
```objc
++i
```
___
*[systems/trajectory_optimization/multiple_shooting.cc, line 157 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16847#-MysYfiRB4Oy8zIpU6i8:-MysYfiRB4Oy8zIpU6i9:b-blgg9u) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa8fac1023549297303867c09c82e6095865289/systems/trajectory_optimization/multiple_shooting.cc#L157)):*
> ```cpp
>     std::vector<solvers::Binding<solvers::Constraint>> b =
>         AddConstraintToAllKnotPoints(f[i]);
>     bindings.insert(bindings.end(), b.begin(), b.end());
> ```

FYI, in case you want to move the elements in `b` into `bindings`, you could do
`bindings.insert(bindings.end(), std::make_move_iterator(b.begin()), std::make_move_iterator(b.end()));`
___
*[systems/trajectory_optimization/sequential_expression_manager.h, line 86 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16847#-Mys_YpkC9e2-S8NwHd5:-Mys_YpkC9e2-S8NwHd6:bmp2djn) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa8fac1023549297303867c09c82e6095865289/systems/trajectory_optimization/sequential_expression_manager.h#L86)):*
> ```objc
>       int index) const;
> 
>   VectorX<symbolic::Variable> GetVariables(
> ```

Missing documentation.
___
*[systems/trajectory_optimization/sequential_expression_manager.cc, line 67 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16847#-Mysc78e2uHl7JakrNp_:-Mysc78fAjU5M3vzBrtO:b2bj48t) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa8fac1023549297303867c09c82e6095865289/systems/trajectory_optimization/sequential_expression_manager.cc#L67)):*
> ```cpp
>   for (int i = 0; i < vars.size(); ++i) {
>     const auto it = expressions_.find(vars[i]);
>     if (it == expressions_.end()) {
> ```

nit, this code path not covered by tests.
___
*[systems/trajectory_optimization/sequential_expression_manager.cc, line 70 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16847#-MysZlNh7SBR3MGncH65:-MysZlNh7SBR3MGncH66:bfpign8) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa8fac1023549297303867c09c82e6095865289/systems/trajectory_optimization/sequential_expression_manager.cc#L70)):*
> ```cpp
>       throw std::runtime_error(
>           vars[i].get_name() +
>           "does not appear to be a placeholder variable in this program.");
> ```

btw you may want a leading space in front of "does not ..."

**Review by xuchenhan-tri — COMMENTED** [review]
+(release notes: yes) 

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16847#-:-MysriAADQBYEIUM5pCM:b-c7f1bi)* status: 5 unresolved discussions (waiting on @RussTedrake)

**Review by hongkai-dai — COMMENTED** [review]
> Reviewed 2 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16847)* status: 3 unresolved discussions (waiting on @aykut-tri, @hongkai-dai, @RussTedrake, and @xuchenhan-tri)

___
*[systems/trajectory_optimization/sequential_expression_manager.h, line 119 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16847#-MysO80w9_Yw9yAMoByy:-Mytiha49ZmtcFOYbGyK:burkoet) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa8fac1023549297303867c09c82e6095865289/systems/trajectory_optimization/sequential_expression_manager.h#L119)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Done.  Good catch.  Note that I've gone ahead and changed the /** */ used throughout this file to /* */ since it is an internal class, and should not use the public doxygen comment blocks.
</blockquote></details>

Not changed yet?

**Review by RussTedrake — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16847)* status: 3 unresolved discussions (waiting on @aykut-tri, @hongkai-dai, and @xuchenhan-tri)

___
*[systems/trajectory_optimization/sequential_expression_manager.h, line 86 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16847#-Mys_YpkC9e2-S8NwHd5:-MyvswGqE9EB1MELP1Ym:bugwife) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa8fac1023549297303867c09c82e6095865289/systems/trajectory_optimization/sequential_expression_manager.h#L86)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.  Good catch.  Note that I've gone ahead and changed the /** / used throughout this file to / */ since it is an internal class, and should not use the public doxygen comment blocks.
___
*[systems/trajectory_optimization/sequential_expression_manager.h, line 119 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16847#-MysO80w9_Yw9yAMoByy:-Myvsxj-3skofUjxZzXV:b-rx9aos) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa8fac1023549297303867c09c82e6095865289/systems/trajectory_optimization/sequential_expression_manager.h#L119)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Not changed yet?
</blockquote></details>

Done.  I accidentally wrote this comment (which was intended for xuchen's review comment above), and missed this one completely!

**Review by xuchenhan-tri — COMMENTED** [review]
> Reviewed 3 of 5 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16847)* status: 2 unresolved discussions (waiting on @aykut-tri and @hongkai-dai)

**Comment by RussTedrake** [comment]
___
*[systems/trajectory_optimization/multiple_shooting.h, line 307 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16847#-MyqaQr-pdX_s0pplZaa-r1-307:-MywgyiP5axOwHd3VQnc:bp90fl8) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa8fac1023549297303867c09c82e6095865289/systems/trajectory_optimization/multiple_shooting.h#L307)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

@aykut-tri -- I need you to OK this comment to clear the PR for merging.

**Comment by aykut-tri** [comment]
___
*[systems/trajectory_optimization/multiple_shooting.h, line 307 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16847#-MyqaQr-pdX_s0pplZaa-r1-307:-MywhMVP1sx3C1QTO2Hh:b-uq8r7l) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa8fac1023549297303867c09c82e6095865289/systems/trajectory_optimization/multiple_shooting.h#L307)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

@aykut-tri -- I need you to OK this comment to clear the PR for merging.
</blockquote></details>

oh sorry about that

**Pull request merged**